modules/developers-guide/pages/graphql-oriented-using.adoc

```diff
@@ -0,0 +1,1062 @@
+UDTs are optional, but if you wish to use a UDT in another object type definition,
+you'll want to create the UDT first.
+Once created, you can include the type in the schema deployed via the GraphQL playground.
+Here are two examples that create a UDT called `Address` that includes a street, city, state, and zipcode,
+and a UDT called `Review` that includes a book title, comment, rating and review data.
+
+[tabs]
+====
+Address UDT::
++
+--
+[source,shell]
+----
+include::example$graphql/1createUDTAddress.graphql[]
+----
+--
+
+Review UDT::
++
+--
+[source,shell]
+----
+include::example$graphql/1createUDTReview.graphql[]
+----
+--
+====
+
+// USE asciidoc callouts to explain? a la https://docs.couchbase.com/home/contribute/code-blocks.html#callouts
+These types have some Cassandra-specific features, in addition to the typical format
+of a GraphQL type.
+To denote that the type will be stored as a UDT in Cassandra, `@cql_entity(target: UDT)`
+is added to the initial line.
+
+To avoid duplicating the type as both an input type and output type, `@cql_input` is
+added to make the type an input type.
+
+Standard CamelCase is used to name the fields, but the Cassandra column name can be
+named with an alternative using `@cql_column(name: "book_title")`.
+In Cassandra, CamelCase names must be double-quoted, so this particular field name is
+both GraphQL friendly (bookTitle) and CQL friendly (book_title).
+
+==== Data types
+
+A column's CQL data type is inferred from the GraphQL field type.
+GraphQL's built-in scalar types are mapped:
+
+[options="header",footer"]
+|====
+| GraphQL | CQL
+| ID      | uuid
+| String  | varchar
+| Int     | int
+| Float   | double
+| Boolean | boolean
+|====
+
+In addition, Stargate provides a set of custom scalar types that map directly
+to the CQL types of the same name:
+Uuid, TimeUuid, Inet, Date, Duration, BigInt, Counter, Ascii, Decimal, Varint,
+Float32, Blob, SmallInt, TinyInt, Timestamp, Time.
+
+Sets and lists are defined by custom typed arrays. Maps and tuples are not
+supported, but UDTs can be used to create the same functionality.
+
+// end::CreateUDT[]
+
+// tag::CreateObjectTypes[]
+=== Creating object types
+
+The most basic components of a GraphQL schema are object types, which just
+represent a kind of object you can fetch from your service, and specify the
+fields contained in the object.
+
+[tabs]
+====
+Book type::
++
+--
+[source,shell]
+----
+include::example$graphql/1createTableBook.graphql[]
+----
+--
+
+Reader type::
++
+--
+[source,shell]
+----
+include::example$graphql/1createTableReader.graphql[]
+----
+--
+====
+
+Although the types defined for UDTs and these examples are not fundamentally different,
+a few new features are illustrated here.
+
+In the Book type, there are additional `@cql_column` parameters defined.
+It is possible to define the partition key and clustering keys are shown, by
+using `partitionKey: true` and `clusteringOrder: ASC`, respectively.
```

Comment:
Maybe worth mentioning that if no fields have `partitionKey: true`, but the first field has type ID, Uuid or TimeUuid, then that field is used as the partition key.

Reply:
changed